### PR TITLE
ICD: Monitoring table, KVS keys minimized.

### LINKED ISSUE #24288: [ICD]  Refactor peristance key used to minimise necessary entries for the Client Monitoring Cluster
### Feature description

Current Implementation requires 1 key entry per fabric to store persistant data for the Client Monitoring Cluster.
Refactor key logic to use 1 key to store the entire Client Monitoring Registration Table.

### Platform

all

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
* Now one KVS key is used per fabric.
* Fixes [#24288](https://github.com/project-chip/connectedhomeip/issues/24288).
ICD Monitoring table and server tests ran successfully.

### CODE REVIEW

File: src/lib/support/PersistentArray.h [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+    PersistentArray(PersistentStorageDelegate * storage) : PersistentData<kMaxSerializedSize>(storage), mLimit(kMaxListSize) {}
+
+    size_t Limit() { return this->mLimit; }
+    uint16_t Count() { return this->mCount; }
+    virtual void Clear(EntryType & entry)                                                = 0;
+    virtual CHIP_ERROR Serialize(TLV::TLVWriter & writer, const EntryType & entry) const = 0;
+    virtual CHIP_ERROR Deserialize(TLV::TLVReader & reader, EntryType & entry)           = 0;
+
+    virtual CHIP_ERROR Copy(EntryType & target, const EntryType & source)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:42]: Again, this needs a lot of documentation about what it's meant to do.  I don't think the way we use this below is safe.

File: src/lib/support/PersistentArray.h [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    {
+        target = source;
+        return CHIP_NO_ERROR;
+    }
+
+    virtual bool Compare(const EntryType & a, const EntryType & b) const { return a == b; }
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:48]: Why is this a virtual method on this class, not a method on EntryType?

File: src/lib/support/PersistentArray.h [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+    void Clear() override
+    {
+        // Clear entries
+        for (size_t i = 0; i < this->mLimit; ++i)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:53]: Why is this going to mLimit, not to mCount?  At the very least, document.

File: src/lib/support/PersistentArray.h [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+#include <lib/support/PersistentData.h>
+
+namespace chip {
+
+enum class PersistentTags : uint8_t
+{
+    kCount   = 1,
+    kList    = 2,
+    kEntries = 3,
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:27]: Where is kEntries used?

File: src/app/icd/IcdMonitoringTable.cpp [hunk 5]
```diff
@@ -5,32 +5,51 @@ namespace chip {
 
 enum class Fields : uint8_t
 {
-    kCheckInNodeID    = 1,
-    kMonitoredSubject = 2,
-    kKey              = 3,
+    kCheckInNodeID = static_cast<uint8_t>(PersistentTags::kEntries),
```
Comment (bzbarsky-apple) [inline src/app/icd/IcdMonitoringTable.cpp:8]: Wait, why are we using the same tag value as someone else?

And do these tag values actually need to be distinct from the ones for PersistentArray?

File: src/lib/support/PersistentArray.h [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        this->mCount = 0;
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, container));
+
+        // Count
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(PersistentTags::kCount), static_cast<uint16_t>(this->mCount)));
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:67]: Why the static_cast here?

File: src/lib/support/PersistentArray.h [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+        // Entries
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(PersistentTags::kList), TLV::kTLVType_List, list));
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:70]: kTLVType_List or kTLVType_Array?  Why is this a list?

File: src/lib/support/PersistentArray.h [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:71]: So wait, can mCount be > mLimit?  Because that's the only reason this condition would look like that...

File: src/lib/support/PersistentArray.h [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        }
+        ReturnErrorOnFailure(writer.EndContainer(list));
+
+        return writer.EndContainer(container);
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+        VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:83]: Why not pass your desired type to Next(), then, and have the TLV machinery do that check for you instead of doing it manually?

File: src/lib/support/PersistentArray.h [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        // Count
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kCount)));
+        ReturnErrorOnFailure(reader.Get(this->mCount));
+
+        // Entries
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kList)));
+        VerifyOrReturnError(TLV::kTLVType_List == reader.GetType(), CHIP_ERROR_INTERNAL);
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:94]: Again, why not just have Next() do the work?

File: src/lib/support/PersistentArray.h [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            ReturnErrorOnFailure(Deserialize(reader, mEntries[i]));
+        }
+        ReturnErrorOnFailure(reader.ExitContainer(list));
+
+        return reader.ExitContainer(container);
+    }
+
+    /**
+     * Stores an entry at a given position. This method does not validates the order
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:107]:
```suggestion
     * Stores an entry at a given position. This method does not validate the order
```

File: src/lib/support/PersistentArray.h [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (this->Compare(e, value))
+            {
+                // Existing value, index must match
+                VerifyOrReturnError(i == index, CHIP_ERROR_DUPLICATE_KEY_ID);
+                ReturnErrorOnFailure(this->Copy(e, value));
+                return this->Save();
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:133]: So this case does not call OnEntryRemoved or OnEntryAdded, right?

Why do we have those hooks?  Is anything using them?

File: src/lib/support/PersistentArray.h [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    /**
+     * @brief Returns the stored entry at a given index.
+     * @param index Zero-based position within the table.
+     * @param entry On success, contains the value matching the given index.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Get(size_t index, EntryType & value)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:198]: This needs to clearly document that the out param value is only valid until the next Set or something.  But it depends on EntryType....

File: src/lib/support/PersistentArray.h [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+    {
+        ReturnErrorOnFailure(this->Load());
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:201]: Again, shouldn't we just ensure that mCount <= mLimit?

File: src/lib/support/PersistentArray.h [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+    /**
+     * @brief Returns the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param index Zero-based position within the table.
+     * @param entry Used as an input for comparison, and output for the rest of the values.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Get(EntryType & value, bool do_create = false)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:215]: So here `value` is an inout parameter, right?  It needs to be _very_ clearly documented in terms of what fields need to be set to what on the "in" side and what state it will put things in when do_create is true.

File: src/lib/support/PersistentArray.h [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+        VerifyOrReturnError(index < this->mCount, CHIP_ERROR_NOT_FOUND);
+        ReturnErrorOnFailure(this->Copy(value, mEntries[index]));
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Returns the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param index Zero-based position within the table.
+     * @param entry Used as an input for comparison, and output for the rest of the values.
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:211]: These docs have nothing to do with the signature of this method.

File: src/lib/support/PersistentArray.h [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+     *        shifted one position down.
+     * @param index Zero-based position within the table.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Remove(size_t index)
+    {
+        ReturnErrorOnFailure(this->Load());
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
+        VerifyOrReturnError(index < this->mCount, CHIP_ERROR_NOT_FOUND);
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:248]: Again, we should just enforce mCount <= mLimit.

File: src/lib/support/PersistentArray.h [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+        EntryType old = mEntries[index];
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:249]: So... this is using `operator=`, not Copy.  Should be documented why?  It's really not clear to me when Copy should be used or not used.

Maybe Copy needs a more descriptive name?

File: src/lib/support/PersistentArray.h [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    /**
+     * @brief Removes the value matching the given entry. The Compare() method
+     *        is used to identify the matching entry within the table.
+     * @param entry Identifies the value to be removed.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if no entry matches the given input.
+     */
+    CHIP_ERROR Remove(EntryType & value)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:269]: Why is the incoming thing non-const?

File: src/lib/support/PersistentArray.h [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Removes all the entries in the table.
+     * @return CHIP_NO_ERROR on success.
+     */
+    CHIP_ERROR RemoveAll()
+    {
+        for (uint16_t i = 0; i < mCount; ++i)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:288]: Don't we need to Load() somewhere here?  Otherwise mCount is not going to be set...  This should be covered by some sort of test, I would think.

File: src/lib/support/PersistentArray.h [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            ReturnErrorOnFailure(Remove(i));
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:290]: This is broken (will remove every other thing, not everything), and presumably nothing tests any codepath with RemoveAll in it for correctness...

File: src/lib/support/PersistentData.h [hunk 22]
```diff
@@ -49,13 +49,23 @@ struct PersistentData
         TLV::TLVWriter writer;
         writer.Init(mBuffer, sizeof(mBuffer));
 
-        ReturnErrorOnFailure(Serialize(writer));
-
-        // Save serialized data
-        return storage->SyncSetKeyValue(key.KeyName(), mBuffer, static_cast<uint16_t>(writer.GetLengthWritten()));
+        CHIP_ERROR err = Serialize(writer);
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentData.h:52]: Why these changes?  This seems logically equivalent to what we had before....

File: src/lib/support/PersistentData.h [hunk 23]
```diff
@@ -49,13 +49,23 @@ struct PersistentData
...
+        if (CHIP_NO_ERROR == err)
+        {
+            // Save serialized data
+            err = storage->SyncSetKeyValue(key.KeyName(), mBuffer, static_cast<uint16_t>(writer.GetLengthWritten()));
+        }
+        return err;
     }
 
-    virtual CHIP_ERROR Load() { return this->Load(this->mStorage); }
+    virtual CHIP_ERROR Load(bool force = false)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentData.h:61]: So the fact that Load() by default does not force means ops on PersistentArray with an entry type that does not own is own data (the only kind we have so far) will lead to a mix of entries pointing into our TLV buffer and entries pointing to caller-provided buffers, no?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/icd/IcdMonitoringTable.cpp [hunk 24]
```diff
@@ -5,32 +5,51 @@ namespace chip {
...
+CHIP_ERROR IcdMonitoringTable::Copy(IcdMonitoringEntry & target, const IcdMonitoringEntry & source)
+{
+    VerifyOrReturnError(kUndefinedNodeId != source.checkInNodeID, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(kUndefinedNodeId != source.monitoredSubject, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(source.key.size() == IcdMonitoringEntry::kKeyMaxSize, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(nullptr != source.key.data(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    target.checkInNodeID    = source.checkInNodeID;
+    target.monitoredSubject = source.monitoredSubject;
+    target.key              = source.key;
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/icd/IcdMonitoringTable.cpp:36]: By IcdMonitoringEntry not owning the buffer, I avoid unnecessary copies. The price, of course, is the added risk of using an invalid buffer. I have added a warning here to clarify.

File: src/app/icd/IcdMonitoringTable.h [hunk 25]
```diff
@@ -20,34 +20,34 @@
...
 #include <lib/core/DataModelTypes.h>
 #include <lib/support/CodeUtils.h>
-#include <lib/support/PersistentData.h>
+#include <lib/support/PersistentArray.h>
 #include <stddef.h>
 
 namespace chip {
 
-constexpr size_t kIcdMonitoringBufferSize = 40;
+constexpr size_t kIcdMonitoringEntrySize  = 40;
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/icd/IcdMonitoringTable.h]: Actually, I miss-calculated, and the actual theoretical value is a couple of bytes higher. The real value, however, is a bit smaller, thanks to some encoding optimizations. I have documented the rational behind this number.

File: src/app/icd/IcdMonitoringTable.h [hunk 26]
```diff
@@ -20,34 +20,34 @@
...
-    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override;
-    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override;
-    void Clear() override;
 
-    chip::FabricIndex fabricIndex = static_cast<chip::FabricIndex>(0);
-    chip::NodeId checkInNodeID    = static_cast<chip::NodeId>(0);
-    uint64_t monitoredSubject     = static_cast<uint64_t>(0);
+    bool operator==(const IcdMonitoringEntry & other) const { return (this->checkInNodeID == other.checkInNodeID); }
+
+    // bool IsValid() { return this->checkInNodeID != kUndefinedNodeId && this->fabricIndex != kUndefinedFabricIndex; }
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/icd/IcdMonitoringTable.h]: Not enough coffee? Removed.

File: src/lib/support/PersistentArray.h (hunk 20 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h:291]: We don't. There is a Load() on each individual Remove(). I have added a flag to prevent unnecessary loads on subsequents calls.
That said, it doesn't hurt, and it is clearer to put an explicit Load() before the loop, so I have.

File: src/lib/support/PersistentArray.h [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+namespace chip {
+
+enum class PersistentTags : uint8_t
+{
+    kCount   = 1,
+    kList    = 2,
+    kEntries = 3,
+};
+
+template <size_t kMaxSerializedSize, size_t kMaxListSize, typename EntryType>
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h]: Good idea. I have modify the template parameter to take <list_size, entry_size, entry_type>.

File: src/lib/support/PersistentData.h [hunk 28]
```diff
@@ -73,12 +83,15 @@ struct PersistentData
         uint16_t size  = static_cast<uint16_t>(sizeof(mBuffer));
         CHIP_ERROR err = storage->SyncGetKeyValue(key.KeyName(), mBuffer, size);
         VerifyOrReturnError(CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND != err, CHIP_ERROR_NOT_FOUND);
-        ReturnErrorOnFailure(err);
-
-        // Decode serialized data
-        TLV::TLVReader reader;
-        reader.Init(mBuffer, size);
-        return Deserialize(reader);
+        if (CHIP_NO_ERROR == err)
```
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/lib/support/PersistentData.h:86]: Makes for easier debugging. Reverted, I can live without this change.

File: src/lib/support/PersistentData.h (hunk 22 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentData.h:52]: No functional change here. I have found myself making this change every time I need to debug these objects. Reverted.

File: src/lib/support/PersistentData.h (hunk 23 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentData.h:57]: Yes. In the case of PersistentArray, after a load, all pointers point to the internal buffer. Those pointer may be modified to point to caller-provider buffers. However, this behavior holds regardless the force flag, which is used to prevent unnecessary NVM reads.
I have changed the default to true, which I think is expected when calling this method directly without argument.

File: src/app/icd/IcdMonitoringTable.h (hunk 25 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/icd/IcdMonitoringTable.h]: I guess the real question, @rcasallas-silabs, is why we are not using the existing TLV::EstimateStructOverhead bits here.

File: src/lib/support/PersistentArray.h (hunk 20 above)
Comment (bzbarsky-apple) [reply src/lib/support/PersistentArray.h]: > There is a Load() on each individual Remove()

That does not help if we are not loaded and mCount is 0, right?  So we really do need the load up front.

File: src/lib/support/PersistentArray.h (hunk 14 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h:207]: We should, and we do, in the Set method. The Get method do not modify either mCount, or mLimit, so checking mCount <= mLimit here is gratuitous.

File: src/lib/support/PersistentArray.h (hunk 15 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h:221]: I have improved the documentation on this function.

File: src/lib/support/PersistentArray.h (hunk 13 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h:207]: Documentation added.

File: src/lib/support/PersistentArray.h (hunk 18 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h:264]: This is an oversight. Copy should be use in all cases to ensure consistency.

File: src/lib/support/PersistentArray.h (hunk 19 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h]: Widespread use of leaded gasoline until 2005, maybe. Const added.

File: src/lib/support/PersistentArray.h (hunk 20 above)
Comment (rcasallas-silabs) [reply src/lib/support/PersistentArray.h]: If not loaded, then it would load on first Remove. After that, if mCount is 0, then there is nothing to remove.
But again, I agree adding the Load here is the way to go.

File: src/app/icd/IcdMonitoringTable.h (hunk 25 above)
[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/icd/IcdMonitoringTable.h]: I guess the real answer is: I didn't know about TLV::EstimateStructOverhead. I would have used, if you mention it on your firs comment.

[the remark this answers is not in the document]
Comment (rcasallas-silabs) [reply src/app/icd/IcdMonitoringTable.h]: I took a look at TLV::EstimateStructOverhead, and it only offers a rough estimate, which won't take in account the overhead of the complex structure required on this case. I stand by my estimate, which is much closer to reality, less wasteful, and now fully documented.

**Review by mkardous-silabs — APPROVED** [review]

File: src/lib/support/PersistentArray.h (hunk 16 above)
Comment (mkardous-silabs) [reply src/lib/support/PersistentArray.h]: @rcasallas-silabs ^

File: src/lib/support/PersistentArray.h [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        // Count
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kCount)));
+        ReturnErrorOnFailure(reader.Get(this->mCount));
+
+        // Entries
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::ContextTag(PersistentTags::kList)));
+        ReturnErrorOnFailure(reader.EnterContainer(list));
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:297]: If we stop because we hit mLimit here, we will have mCount > mLimit, which seems odd.

Maybe it would be better to clamp mCount to be <= mLimit after we `Get(this->mCount)` instead?

File: src/lib/support/PersistentArray.h [hunk 30]
```diff
@@ -0,0 +1,317 @@
...
+
+        return reader.ExitContainer(container);
+    }
+
+    virtual void OnEntryAdded(const EntryType & entry) {}
+    virtual void OnEntryRemoved(const EntryType & entry) {}
+    virtual void OnEntryModified(const EntryType & old_value, const EntryType & new_value) {}
+
+private:
+    // EntryType & At(size_t index) { return (index < this->mLimit) ? mEntries[index] : mEntries[0]; }
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:311]:
```suggestion
```

File: src/lib/support/PersistentArray.h [hunk 31]
```diff
@@ -0,0 +1,317 @@
...
+    EntryType mEntries[kMaxListSize];
+    const size_t mLimit = 0;
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:313]: We end up setting mLimit to kMaxListSize in the constructor.... but why not do it here?  kMaxListSize is a template parameter, so a compile-time known constant....

More to the point, why do we need to store it in the member at all if it's a compile constant attached to the class?

File: src/lib/support/PersistentArray.h [hunk 32]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    /**
+     * @brief Removes all the entries in the table.
+     * @return CHIP_NO_ERROR on success.
+     */
+    CHIP_ERROR RemoveAll()
+    {
+        ReturnErrorOnFailure(this->Load());
+        for (uint16_t i = 0; i < mLimit; ++i)
```
Comment (bzbarsky-apple) [inline src/lib/support/PersistentArray.h:243]: Why is this going to mLimit (which includes un-initialized data, right?) instead of just mCount?
